src/ortho.rs

```diff
@@ -1,71 +1,49 @@
+use std::collections::HashMap;
+use std::sync::{Mutex, OnceLock};
+
 #[derive(Debug, Clone)]
 pub struct Ortho {
     version: u64,
     storage: Vec<u16>,
     dimensions: Vec<u16>,
 }
 
-impl Ortho {
-    pub fn new(version: u64) -> Self {
-        // Use minimum dimensions [2,2] for orthogonality
-        Ortho { 
-            version,
-            storage: Vec::new(),
-            dimensions: vec![2, 2],
-        }
-    }
-    
-    pub fn with_dimensions(version: u64, dimensions: Vec<u16>) -> Self {
-        Ortho {
-            version,
-            storage: Vec::new(),
-            dimensions,
-        }
-    }
+// Global cache for logical coordinates keyed by dimensions
+static LOGICAL_COORDINATES_CACHE: OnceLock<Mutex<HashMap<Vec<u16>, Vec<Vec<u16>>>>> = OnceLock::new();
 
-    pub fn version(&self) -> u64 {
-        self.version
-    }
+// Global cache for forbidden values keyed by (dimensions, storage_length)
+static FORBIDDEN_CACHE: OnceLock<Mutex<HashMap<(Vec<u16>, usize), Vec<u16>>>> = OnceLock::new();
+
+// Global cache for required values keyed by (dimensions, storage)
+static REQUIRED_CACHE: OnceLock<Mutex<HashMap<(Vec<u16>, Vec<u16>), Vec<Vec<u16>>>>> = OnceLock::new();
```

Comment:
Do not key anything by storage. Storage is heavily variant while shape, length, and position are only lightly variant